core/src/libraries/Rewards.sol

```diff
@@ -0,0 +1,122 @@
+     * @dev Use `Rewards.pre()` to easily obtain the first two arguments
+     * @param store The rewards storage pointer
+     * @param accumulated Up-to-date `poolState.accumulated`, i.e. the output of `_accumulate`
+     * @param oldTotalSupply The `totalSupply` before mint/burn
+     * @param newTotalSupply The `totalSupply` after mint/burn
+     */
+    function updatePoolState(
+        Storage storage store,
+        uint168 accumulated,
+        uint256 oldTotalSupply,
+        uint256 newTotalSupply
+    ) internal {
+        PoolState memory poolState = store.poolState;
+
+        poolState.accumulated = accumulated;
+        poolState.lastUpdated = uint32(block.timestamp);
+        poolState.rate = _rate(poolState.rate, oldTotalSupply, newTotalSupply);
+
+        store.poolState = poolState;
+    }
+
+    /**
+     * @notice Tracks how much reward a `user` earned while holding a particular `balance`. Should be
+     * called anytime their balance changes.
+     * @dev Use `Rewards.pre()` to easily obtain the first two arguments
+     * @param store The rewards storage pointer
+     * @param accumulated Up-to-date `poolState.accumulated`, i.e. the output of `_accumulate`
+     * @param user The user whose balance (# of shares) is about to change
+     * @param balance The user's balance (# of shares) -- before it changes
+     */
+    function updateUserState(Storage storage store, uint168 accumulated, address user, uint256 balance) internal {
+        unchecked {
+            UserState memory userState = store.userStates[user];
+
+            userState.earned += uint88((balance * (accumulated - userState.checkpoint)) / 1e18);
+            userState.checkpoint = accumulated;
+
+            store.userStates[user] = userState;
+        }
+    }
+
+    /// @dev Returns arguments to be used in `updatePoolState` and `updateUserState`. No good semantic
+    /// meaning here, just a coincidence that both functions need this information.
+    function pre() internal view returns (Storage storage store, uint168 accumulator) {
+        store = _getStorage();
+        accumulator = _accumulate(store.poolState);
+    }
+
+    /// @dev Accumulates rewards based on the current `rate` and time elapsed since last update
+    function _accumulate(PoolState memory poolState) private view returns (uint168) {
+        unchecked {
+            uint256 deltaT = block.timestamp - poolState.lastUpdated;
+            return poolState.accumulated + uint168(poolState.rate * deltaT);
+        }
+    }
+
+    /// @dev Adjusts `rate` to account for changes in `totalSupply`
+    function _rate(uint56 rate, uint256 oldTotalSupply, uint256 newTotalSupply) private pure returns (uint56) {
+        unchecked {
+            if (oldTotalSupply == 0) oldTotalSupply = 1;
+            if (newTotalSupply == 0) newTotalSupply = 1;
+
+            return uint56((rate * oldTotalSupply) / newTotalSupply);
```

Comment:
Does it make sense to return a non-zero rate if there is no supply?

Reply:
Also, wouldn't the rate be disproportionate if the total supply is zero and new tokens are minted?